### ISSUE TITLE
windows shell support?

### ISSUE THREAD

**Issue by faridcher** [r-lib/crayon#62]
Please add support for powershell or cmd (+[ansicont](ansicon.adoxa.vze.com)] 
Now, I can output colored text in powershell (cmd+ansicon) using:

```r
cat("\033[32;1m OK \033[0m\n")
```
but crayon functions don't work!

**Comment by gaborcsardi** [comment]
Note that you can force-enable crayon with the `crayon.enabled` and the `crayon.colors` options, as a workaround.

Can can we detect color support in powershell or with ansicont?

**Comment by faridcher** [comment]
run powershell then `rterm.exe` and paste this:

> cat("\033[32;1m OK \033[0m\n")

you will get colored output....

**Comment by gaborcsardi** [comment]
Yeah, but is there a way to detect this programmatically? How does R know that it is runnint in powershell, or in rterm with color support?

**Comment by faridcher** [comment]
No idea! Is it important (it might be important if you want to show the user a meaningful message!)? the above line of code proves the support. If `crayon` does what it is supposed to do, then we should get colored output!

**Comment by gaborcsardi** [comment]
Crayon cannot emit color sequences, unless it knows that the platform supports it. 

Again, if you just want to make it work for yourself, run this, maybe from your `.Rprofile`:
```r
options(crayon.enabled = TRUE)
options(crayon.colors = 256)
```

**Comment by wlandau-lilly** [comment]
What about enabling crayon in RStudio? I tried setting just `options(crayon.enabled = TRUE)` and then tried setting both options, but the color characters print out as text: "[38;5;105m", "[38;5;32m", etc. Should I try changing the encoding? I cannot find an encoding setting in RStudio's global options. For what it's worth, I also had a look at the README and #56.

**Comment by wlandau-lilly** [comment]
I also submitted an [SO post](https://stackoverflow.com/questions/46429848/enabling-crayon-in-the-rstudio-console)

**Comment by wlandau-lilly** [comment]
Confirmed: the [RStudio 1.1 preview](https://www.rstudio.com/products/rstudio/download/preview/) supports `crayon` colors automatically. I had no idea.

**Comment by sthupakula-rms** [comment]
> What about enabling crayon in RStudio? I tried setting just `options(crayon.enabled = TRUE)` and then tried setting both options, but the color characters print out as text: "�[38;5;105m", "�[38;5;32m", etc. Should I try changing the encoding? I cannot find an encoding setting in RStudio's global options. For what it's worth, I also had a look at the README and #56.

It's an old post, but did you find a fix for this? I still can't get colors on my DOS window, even with both the options set.

**Comment by gaborcsardi** [comment]
It  appears that  your DOS window does not support  ANSI colors.

**Comment by sthupakula-rms** [comment]
Thank you. I found the solution [here](https://github.com/adoxa/ansicon/blob/master/readme.txt): 
This can be used within R script by calling the exe like this
`system("D:/completePath/x64/ansicon.exe -p")`

and then using regular crayon code! Happier than ever!

**Comment by gaborcsardi** [comment]
I wonder if we could somehow autodetect when ansicon is running.

**Comment by ffollett** [comment]
I believe I'm still experiencing this issue (or similar).

I'm in RStudio, R version 3.6.1, Platform: x86_64-w64-mingw32/x64

```
>packageVersion("crayon")
[1] '1.3.4'
>green("Hello World")
[1] "\033[32mHello World\033[39m"
```

This is a new development for me.  I had crayon working correctly in a script I'm developing, but I did a fresh install yesterday and now it doesn't function correctly.

**Comment by ffollett** [comment]
I think I'm experiencing something similar.  I had crayon working properly in RStudio, but did a fresh install of Windows yesterday (and of R and RStudio) and now it's broken.

It works properly in the console, but when I save what I've done in a console to a script and run it in the RStudio terminal (set to PowerShell), it doesn't color the text and just shows me the markup or whatever that is:

```
#test.R
library(crayon)
packageVersion("crayon")
cat(green("Hello World"),"\n")
```

```
> Rscript test.R
[1] '1.3.4'
?[32mHello World?[39m
```

R Version 3.6.1, R Studio v1.2.1335, Platform: x86_64-w64-mingw32/x64 (64-bit)

**Comment by gaborcsardi** [comment]
Apparently, your powershell does not support ANSI colors.

**Comment by ffollett** [comment]
It definitely supports coloring, but it looks like it uses different escape sequences.  More info [here](https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#text-formatting).
![image](https://user-images.githubusercontent.com/7415274/62234610-a36cf480-b390-11e9-882c-89f6af69c5b6.png)
Perhaps as mentioned above, if you could programaticaly detect Windows Powershell, crayon could use an alternate escape sequence?

**Comment by gaborcsardi** [comment]
No, the `\033` is the same as `$([char]27)`.

**Comment by ffollett** [comment]
That's generally true, but windows has to be different.  Powershell doesn't use `\` as an excape character (because windows uses it for file paths), so \033 is just interpreted as a char string, not the octal bash understands it as.  Hence, you have to embed an expression, such as $([char]27).

![image](https://user-images.githubusercontent.com/7415274/62249300-b17e3d80-b3af-11e9-9dd9-ea4c09b5487e.png)

**Comment by gaborcsardi** [comment]
`\033` is interpreted by R already.

**Comment by 2005m** [comment]
Color in a windows shell is definitely possible. Here is an example: 
https://stackoverflow.com/questions/9203362/c-color-text-in-terminal-applications-in-windows
Git and GCC have colored output in windows terminal. As usual some just need to do it...I might give it a try if I am bored.
